src/Features/Core/Portable/ConvertForEachToFor/AbstractConvertForEachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,474 @@
+            {
+                return null;
+            }
+
+            var foreachCollection = RemoveImplicitConversion(operation.Collection);
+            if (foreachCollection == null)
+            {
+                return null;
+            }
+
+            GetInterfaceInfo(semanticFact, model, foreachVariable, foreachCollection,
+                out var explicitCastInterface, out var collectionNameSuggestion, out var countName);
+            if (countName == null)
+            {
+                return null;
+            }
+
+            var requireCollectionStatement = CheckRequireCollectionStatement(foreachCollection);
+            return new ForEachInfo(
+                semanticFact, options, collectionNameSuggestion, countName,
+                explicitCastInterface, foreachVariable.Type, requireCollectionStatement, foreachStatement);
+        }
+
+        private static void GetInterfaceInfo(
+            ISemanticFactsService semanticFact, SemanticModel model, ILocalSymbol foreachVariable, IOperation foreachCollection,
+            out ITypeSymbol explicitCastInterface, out string collectionNameSuggestion, out string countName)
+        {
+            explicitCastInterface = default;
+            collectionNameSuggestion = default;
+            countName = default;
+
+            // go through list of types and interfaces to find out right set;
+            var foreachType = foreachVariable.Type;
+            if (IsNullOrErrorType(foreachType))
+            {
+                return;
+            }
+
+            var collectionType = foreachCollection.Type;
+            if (IsNullOrErrorType(collectionType))
+            {
+                return;
+            }
+
+            // go through explicit types first.
+
+            // check array case
+            if (collectionType is IArrayTypeSymbol array && array.Rank == 1)
+            {
+                if (!IsExchangable(semanticFact, array.ElementType, foreachType, model.Compilation))
+                {
+                    return;
+                }
+
+                collectionNameSuggestion = "array";
+                explicitCastInterface = null;
+                countName = Length;
+                return;
+            }
+
+            // check string case
+            if (collectionType.SpecialType == SpecialType.System_String)
+            {
+                var charType = model.Compilation.GetSpecialType(SpecialType.System_Char);
+                if (!IsExchangable(semanticFact, charType, foreachType, model.Compilation))
+                {
+                    return;
+                }
+
+                collectionNameSuggestion = "str";
+                explicitCastInterface = null;
+                countName = Length;
+                return;
+            }
+
+            // check ImmutableArray case 
+            if (collectionType.OriginalDefinition.Equals(model.Compilation.GetTypeByMetadataName(typeof(ImmutableArray<>).FullName)))
```

Comment:
seems weird to have to special case this single type.  Is there a reason that's necessary?  can we not jsut tell this is an applicable type because it has GetEnumerator, as well as indexer and Length?

Reply:
it doesn't check for pattern, it looks for explicit interface to be implemented for the refactoring is offered.

all these will probably changed in next iteration (not this PR), since we set the principle as always offer code transformation. so a lot of case where it doesn't offer refactoring now, will offer refactoring.  such as writtenInto, next variable, missing local variable or collection, error type and etc.

Reply:
Sorry... why would we need multiple PRs here.  Now that we have more of a sense of what the principles are, why not just apply them to htis PR?   Is there a need to get this in in one state, only to fix it up with new principles immediately afterwards?

Reply:
I will tagging @jinujoseph here. since those are decided to be done later.